Documentation/design-docs/raw-eventlistener.md

```diff
@@ -0,0 +1,81 @@
+
+A raw API is one that dispatches events, but does not decode the data payload.  It simply provides a raw blob and a metadata blob that can be used to interpret the data when desired.
+
+## API Surface ##
+
+The existing EventListener API usage pattern is to:
+
+1. Create a new class that derives from EventListener.
+2. Implement the OnEventSourceCreated method to get callbacks when new EventSource objects are created (this allows EventListeners to subscribe to events).
+3. Implement the OnEventWritten method to get callbacks whenever a subscribed event is dispatched.
+
+The proposed extension of this pattern is the following:
+
+1. Add a new EventListener constructor that takes an EventListenerSettings enum parameter:
+
+```
+[Flags]
+public enum EventListenerSettings
+{
+   None,
+   RawEventDispatch
+}
+``` 
+
+This parameter is used to specify the desired dispatch behavior (in this case, do not deserialize event payloads).
+
+2. Depending on the configuration, EventListener will call the traditional (deserialized payload) API or the raw API.
+
+The new raw dispatch API will be:
+
+```
+public void OnEventWrittenRaw(RawEventWrittenEventArgs args);
+ 
+public sealed class RawEventWrittenEventArgs
+{
+
+    // Event metadata copied from EventWrittenEventArgs (for consistency).
+    public string EventName { get; }
+    public int EventId { get; }
+    public Guid ActivityId { get; }
+    public Guid RelatedActivityId { get; }
+    public EventSource EventSource { get; }
+    public EventKeywords Keywords { get; }
+    public EventOpcode Opcode { get; }
+    public EventTask Task { get; }
+    public EventTags Tags { get; }
+    public string Message { get; }
+    public byte Version { get; }
+    public EventLevel Level { get; }
+    public long OSThreadId { get; }
+    public DateTime TimeStamp { get; }
+    
+    // Replacement properties for Payload and PayloadNames.
+    public ReadOnlySpan<byte> Metadata { get; }     
+    public ReadOnlySpan<byte> Payload { get; }
+}
+```
+
+The lifetime of the RawEventWrittenEventArgs object will be the lifetime of the callback.  Thus, we can recycle the object and reduce the overhead of this API.  This is similar to how TraceEvent works today, but is different than the existing EventListener.OnEventWritten callback.
+
+NOTE: Until the API becomes public, testing can be performed by reflection using the following steps:
+
+1. Set the configuration flags via private reflection.
+2. Registering a callback that matches the required signature of the raw API using private reflection.
+
+It is possible that due to this reflection requirement, some calls to the deserialized payload API will be made until EventListener discovers that this is no longer needed.  This will occur if setting the flag from the constructor.  This can be worked around by setting the flag from OnEventSourceCreated prior to calling EnableEvents.
```

Comment:
Yeah, the "race" (it's not so much a race, since you always lose ;)) between the derived constructor and `OnEventSourceCreated` is one of our biggest frustrations with `EventListener` 😄. So we're familiar with tricks like this.